### PR TITLE
feat: monorepo

### PR DESCRIPTION
**Pull request by vanpho93**
Resolves: #6389
Impact: minor
Type: feature

## Issue

Too many repos.

## Solution

Using changeset and npm workplace to convert things into monorepo

If you're solving a UIX related issue, please attach screen-caps or gifs showing how your solution differs from the issue.

### CODE REVIEW

File: .changeset/config.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "https://unpkg.com/@changesets/config@2.1.1/schema.json",
+  "changelog": ["@changesets/changelog-github", { "repo": "vanpho93/reaction-monorepo" }],
```
Comment (aldeed) [inline .changeset/config.json]: Need to change to correct `repo`

File: .circleci/config.yml [hunk 2]
```diff
@@ -53,7 +59,9 @@ jobs:
             - reaction-v6-node-modules-{{ .Branch }}
       - run:
           name: Lint JavaScript
-          command: npm run lint
+          command: |
+            npm run build:packages
+            npm run lint
```
Comment (aldeed) [inline .circleci/config.yml]: Is building necessary to run eslint and tests? Usually it is not.

File: .circleci/config.yml [hunk 3]
```diff
@@ -125,10 +149,13 @@ jobs:
           keys:
             - reaction-v6-node-modules-{{ checksum "package.json" }}-{{ checksum "package-lock.json" }}
             - reaction-v6-node-modules-{{ .Branch }}
+      - run:
+          name: Authenticate with registry
+          command: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > .npmrc
```
Comment (aldeed) [inline .circleci/config.yml]: This will overwrite the contents of `.npmrc`. You might want to do `> ~/.npmrc` to avoid losing the contents of the repo file, and/or do `>>` instead of `>` to append to the existing file contents.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      - name: Install Dependencies
+        run: npm install
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }} # Ensure to have this set up under GitHub secrets
+      - name: Create Release Pull Request or Publish to npm
+        uses: changesets/action@v1
+        with:
+          publish: npm run release
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # Ensure to have this set up under GitHub secrets
```
Comment (aldeed) [inline .github/workflows/release.yml]: You don't set `GITHUB_TOKEN` in secrets. Instead, you add `permissions` object on this workflow job, and then `GITHUB_TOKEN` will be automatically available with those permissions. Docs: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

File: package.json [hunk 5]
```diff
@@ -1,51 +1,21 @@
 {
-  "name": "reaction-api",
-  "version": "4.2.0",
+  "name": "reaction",
+  "version": "1.0.0",
```
Comment (aldeed) [inline package.json]: Need to decide what version this should be. Or omit the version property.

File: package.json [hunk 6]
```diff
@@ -1,51 +1,21 @@
...
-          ],
-          "message": "chore(release): v${nextRelease.version} [skip ci]"
-        }
-      ]
-    ]
+  "main": "index.js",
+  "scripts": {
+    "start:dev": "npm run start:dev -w=apps/reaction",
+    "start:meteor-blaze-app": "npm run start -w=apps/meteor-blaze-app",
+    "build:packages": "npm run build -w=packages/file-collections -w=packages/file-collections-sa-base -w=packages/file-collections-sa-gridfs -w=packages/logger -w=packages/random",
```
Comment (aldeed) [inline package.json]: I think it would be better to do `npm run build --if-present --ws` so that it will build any new packages automatically.

File: package.json [hunk 7]
```diff
@@ -1,51 +1,21 @@
...
+    "test": "npm run test --if-present --ws",
+    "test:integration:query": "npm run test:integration:query --workspace=apps/reaction",
+    "test:integration:mutation": "npm run test:integration:mutation --workspace=apps/reaction",
+    "changeset": "changeset",
+    "release": "changeset publish",
+    "lint": "eslint . --resolve-plugins-relative-to .",
+    "lint:docker": "npx --quiet --package @reactioncommerce/ci-scripts@1.6.2 lint-dockerfiles --workspace=apps/reaction",
+    "lint:gql": "npm run lint:gql -w=apps/reaction -w=packages/api-core -w=packages/api-plugin-surcharges",
```
Comment (aldeed) [inline package.json]: I think it would be better to do `npm run lint:gql --if-present --ws` so that it will lint GraphQL for all packages that have it set up.

**Review by aldeed — CHANGES_REQUESTED** [review]
Looks pretty good! I left some inline comments, and you can delete .circleci and .github folders from every package/app folder.

File: .github/workflows/release.yml (hunk 4 above)
Comment (vanpho93) [reply .github/workflows/release.yml]: We don't manually set GITHUB_TOKEN in secrets, this is just get the value only.

File: apps/reaction/tests/integration/api/mutations/addTag/addTag.test.js [hunk 8]
```diff
@@ -154,14 +154,11 @@ describe("authorized user", () => {
   });
   test("cannot create tag with existing slug", async () => {
     const logLevel = Logger.level();
-    const duplicateTagSlug = Factory.Tag.makeOne({
```
Comment (brent-hoover) [inline apps/reaction/tests/integration/api/mutations/addTag/addTag.test.js:157]: Confused why this change is here?

File: apps/reaction/tests/integration/api/mutations/updateTag/updateTag.test.js [hunk 9]
```diff
@@ -182,7 +182,7 @@ describe("authorized user", () => {
       slug: "tag-slug"
     });
     Logger.level("FATAL");
-    expect.assertions(1);
+    // expect.assertions(1);
```
Comment (brent-hoover) [inline apps/reaction/tests/integration/api/mutations/updateTag/updateTag.test.js]: Why is this expect removed?

File: packages/api-core/package.json [hunk 10]
```diff
@@ -50,7 +51,7 @@
   "dependencies": {
     "@apollo/federation": "^0.15.0",
     "@graphql-tools/merge": "^8.0.1",
-    "@reactioncommerce/api-utils": "^1.12.0",
```
Comment (brent-hoover) [inline packages/api-core/package.json]: Is this actually what we want here?

File: packages/api-plugin-files/package.json [hunk 11]
```diff
@@ -21,24 +22,22 @@
...
   "dependencies": {
     "@reactioncommerce/api-utils": "^1.16.7",
-    "@reactioncommerce/file-collections": "~0.9.3",
+    "@reactioncommerce/file-collections": "^0.9.3",
     "@reactioncommerce/file-collections-sa-gridfs": "~0.1.5",
     "@reactioncommerce/logger": "~1.1.4",
     "@reactioncommerce/random": "~1.0.2",
     "@reactioncommerce/reaction-error": "~1.0.1",
-    "envalid": "~5.0.0",
+    "envalid": "^7.3.1",
```
Comment (brent-hoover) [inline packages/api-plugin-files/package.json]: We should be keeping the version of envalid consistent at 6.x.x

File: packages/api-plugin-inventory-simple/src/index.js [hunk 12]
```diff
@@ -45,6 +45,7 @@ export default async function register(app) {
     },
     mutations,
     policies,
-    queries
+    queries,
+    additionalFields: { test: 4 }
```
Comment (brent-hoover) [inline packages/api-plugin-inventory-simple/src/index.js]: I think this is leftover test code?

File: packages/api-plugin-pricing-simple/src/index.js [hunk 13]
```diff
@@ -45,6 +45,7 @@ export default async function register(app) {
     },
     simpleSchemas: {
       PriceRange
-    }
+    },
+    additionalFields: { test: 4 }
```
Comment (brent-hoover) [inline packages/api-plugin-pricing-simple/src/index.js]: More test code?

File: packages/api-utils/package.json [hunk 14]
```diff
@@ -70,8 +71,7 @@
     "eslint-plugin-react": "^7.19.0",
     "eslint-plugin-react-hooks": "^3.0.0",
     "graphql": "^15.0.0",
-    "jest": "^26.0.1",
-    "semantic-release": "^17.0.4"
+    "jest": "^26.0.1"
```
Comment (brent-hoover) [inline packages/api-utils/package.json]: Are we keeping the version of this consistent?

**Review by brent-hoover — COMMENTED** [review]
Just have a few questions

File: apps/reaction/tests/integration/api/mutations/addTag/addTag.test.js (hunk 8 above)
Comment (vanpho93) [reply apps/reaction/tests/integration/api/mutations/addTag/addTag.test.js:157]: Fixed by adding shopId. This change caused by an unique index changed from `slug: 1` into `slug: 1, shopId: 1`.

File: packages/api-utils/package.json (hunk 14 above)
Comment (vanpho93) [reply packages/api-utils/package.json]: There are many inconsistent dependencies, I am not sure if we could solve them all in this PR.

File: .circleci/config.yml (hunk 2 above)
Comment (vanpho93) [reply .circleci/config.yml]: image
Without `build:packages` there will be an error. Maybe there's a way around to fix this.

File: packages/api-utils/package.json (hunk 14 above)
Comment (brent-hoover) [reply packages/api-utils/package.json]: Yeah, but we're touching this one in this PR

Comment (brent-hoover) [reply packages/api-utils/package.json]: Sorry, misread the diff. Disregard the comment. But we should do a push to make all common dependencies consistent.

**Comment by tedraykov** [comment]
The Circle CI run doesn't seem to be completely configured. We don't want to run linting and unit testing on every package on every commit, because it is redundant. Most CI/CD tools provide some sort of path filtering that run jobs only for the changed directories. 

Another concern I have is if we want to release a new docker image for `reaction`, we must be sure the integration tests that are running before the `docker-build-push` are using the newly published api-plugins.
Let's say we update the `orders` plugin and we run `npm run changeset` to generate the changesets. The changeset bot will generate a PR that bumps the `orders` plugin version and also update that version in the `reaction` api. When we merge this `release` PR, the circle ci and the github actions runs will start in parallel. A problem that might arise is when circle ci tries to build the `reaction` app that depends on the new `orders` plugin version, that might not be published by the github action job.

File: .github/workflows/release.yml (hunk 4 above)
Comment (aldeed) [reply .github/workflows/release.yml]: I was referring to the comment, which can be removed. And also I didn't see a `permissions` section in the yaml at the time, but maybe the default permissions are fine.

File: packages/api-utils/package.json (hunk 14 above)
Comment (vanpho93) [reply packages/api-utils/package.json]: I updated jest and eslint. i think we can do more later.

**Comment by vanpho93** [comment]
> The Circle CI run doesn't seem to be completely configured. We don't want to run linting and unit testing on every package on every commit, because it is redundant. Most CI/CD tools provide some sort of path filtering that run jobs only for the changed directories.

This's an improvement but we can do it later, there is a (tool)[https://turborepo.org/] that @aldeed suggested which could help.

> Another concern I have is if we want to release a new docker image for reaction, we must be sure the integration tests that are running before the docker-build-push are using the newly published api-plugins.

Integration tests run outside of docker.
There are still problem when a plugin upgrade trigger reaction app upgrade. In that case, `changeset` will update only package.json but not package-lock.json. I am trying to fix this.

@tedraykov @zenweasel

**Review by aldeed — CHANGES_REQUESTED** [review]
I'm not sure about CircleCI, but for GitHub Actions, there is a good action plugin, `technote-space/get-diff-action`, that you can use to check folder diffs before running additional per-project checks.

```yaml
jobs:
  project-changed:
    runs-on: ubuntu-latest
    outputs:
      didChange: ${{ steps.projectDiff.outputs.count > 0 }}
    steps:
      - uses: actions/checkout@v3
      - name: Find changed files
        id: projectDiff
        uses: technote-space/get-diff-action@v6
        with:
          # List files and/or folders that should trigger remaining jobs when changed
          PATTERNS: |
            packages/package/**

  next-job:
    runs-on: ubuntu-latest
    # Skip if no project files have changed
    needs: project-changed
```

Note that it usually isn't correct to use the CI `paths` feature because that only checks the most recent commit, whereas you would want to run jobs if _any_ commit included in a PR has modified files in that folder.

My other concerns are:
- I don't know if shrinkwrap works with all of this. `api-plugin-authentication` is the only package that uses shrinkwrap. It's being used to pin a subdependency to an older version. You should remove the shrinkwrap file and instead use the newer [overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) feature.
- To gain full advantages of the monorepo, you should migrate npm locks to the root package-lock and remove package-lock files from all packages.
- Did you test whether developing locally with multiple layers of package dependencies is working? For example, if running the app, which depends on `file-collections-sa-gridfs`, which in turn depends on `file-collections-sa-base`, can you edit `file-collections-sa-base` with hot reloading?

**Review by aldeed — CHANGES_REQUESTED** [review]
A few more suggestions...

In the root `package.json`, add:

```json
"engines": {
    "node": ">=14.18.1",
    "npm": ">=7"
  },
  "engineStrict": true,
```

To support NPM v8.6+, run the following command and commit ([explanation](https://github.com/npm/cli/issues/4664#issuecomment-1086096726)):

```txt
npm config set legacy-peer-deps=true --location=project
```

(For this repo, there's a conflicting `graphql` version for which peer warnings were previously ignored. Eventually you could update packages to make everything happy and then migrate to strict peer deps.)

When I run `npm run start:dev`, there is a package error `Cannot find module 'json-bigint'`, which is coming from a dependency of `tus-node-server`. You can fix it by updating `tus-node-server` to `"~0.7.1"` in `packages/file-collections/package.json` and then re-run `npm install` in repo root. It seems like there are not any breaking changes with this update, but you should re-test file uploads to be sure. You may also need to update to a newer `tus-js-client`, which might have breaking changes.

In the new section of the readme, you are missing two steps for being able to run the app:
- Copy `apps/reaction/.env.example` to `apps/reaction/.env`.
- Start MongoDB. Maybe you can add a `docker-compose.yml` file in the repo root that just starts mongo, so they can run `docker-compose up -d` before `npm run start:dev`? It would be similar to the `docker-compose.yml` file in `apps/reaction`, but without the app service. You would also have to change the MONGO_URL to `mongodb://localhost:27017/reaction` in the `.env.example` file.

**Comment by brent-hoover** [comment]
If we add a docker-compose with Mongo, we could use the one from the CLI and the CLI could grab it from here

File: README.md [hunk 15]
```diff
@@ -60,6 +60,44 @@ When make completes, three services will be running on localhost:
 
 Go to the complete [installation instructions](https://mailchimp.com/developer/open-commerce/guides/quick-start/#access-the-dashboard-playground-and-storefront) to see how to set up your store
 
+# API development
+
+## What you need
+* NPM >= 7 (support workspaces)
+* Docker & docker-compose
+* MongoDB
```
Comment (brent-hoover) [inline README.md]: This is incorrect, you don't need Mongo installed if you are using the docker-compose and most people are better of not installing it locally so you can have different instances of it

File: .github/workflows/release.yml [hunk 16]
```diff
@@ -12,21 +12,30 @@ jobs:
...
+
       - name: Use Node.js 14.x
         uses: actions/setup-node@v2
         with:
           node-version: 14
 
-      - name: Install npm 8.1.0
-        run: npm i -g npm@8.1.0
+      - name: Install pnpm
+        run: npm i -g pnpm@latest
```
Comment (brent-hoover) [inline .github/workflows/release.yml:31]: I think we had agreed to pin npm at a particular version? There has been problems with it pulling in a breaking version of npm

File: packages/api-plugin-sample-data/package.json [hunk 17]
```diff
@@ -10,6 +10,7 @@
   "homepage": "https://github.com/reactioncommerce/reaction",
   "url": "https://github.com/reactioncommerce/reaction.git",
   "email": "hello-open-commerce@mailchimp.com",
+  "private": true,
```
Comment (brent-hoover) [inline packages/api-plugin-sample-data/package.json:13]: I think this disables publishing to npm, which is something we do want?

File: packages/api-core/src/importPluginsJSONFile.js [hunk 18]
```diff
@@ -41,7 +41,11 @@ export default async function importPluginsJSONFile(pluginsFile, transformPlugin
       Logger.debug({ pluginPath, pluginRefs });
       throw new Error(`Plugin "${name}" is not set to a string`);
     } else if (/[a-zA-Z@]/.test(pluginPath[0])) {
-      ({ default: plugin } = await import(pluginPath));
+      try {
```
Comment (brent-hoover) [inline packages/api-core/src/importPluginsJSONFile.js:44]: What's the reasoning for this?

**Review by brent-hoover — APPROVED** [review]

File: .github/workflows/release.yml [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+      - name: Install Dependencies
+        run: pnpm install --ignore-scripts
+
+      - name: Create Release Pull Request or Publish to npm
+        uses: changesets/action@v1
+        with:
+          publish: pnpm release
```
Comment (tedraykov) [inline .github/workflows/release.yml:39]: I'm skeptical this would work as intended. First of all there is no pnpm command called release, also there are no scripts called `release` in the plugins. When we merge this in trunk, are we sure what part of the repository will be published?

File: Dockerfile [hunk 20]
```diff
@@ -1,5 +1,36 @@
 # Dockerfile for production builds
-FROM reactioncommerce/node-prod:14.18.1-v1
+# syntax=docker/dockerfile:1.4
+
+FROM node:14.18.1-alpine as deps
```
Comment (tedraykov) [inline Dockerfile]: This docker file is better suited for the app folder itself, not In the root of the repo. If you want to build a container image with the root context but use the `apps/reaction/Dockerfile` you can specify this in the `docker build` command like so:
```
docker build -t ... -f ./apps/reaction/Dockerfile .
```

**Review by tedraykov — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 19 above)
Comment (vanpho93) [reply .github/workflows/release.yml:39]: Actually release will be called on circleci on `pnpm publish -r`.
